### PR TITLE
GWAS input handling improvements

### LINKED ISSUE #752: GWAS input handling improvements
- [x] Do not require exact match of samples in geno and pheno inputs for GWAS. Instead use intersection of samples for analysis and inform user about ignored samples.
- [x] Handle HapMap files supported by GAPIT (slightly different headers, content is the same). 
- [x] Show more details if file identified as HapMap could not be parsed.
- [x] Allow to download example of HapMap file.
- [x] Add info about preferred file encoding.

### LINKED ISSUE #772: GWAS results datatable error when no map file is present
In case analysis was performed with CSV as geno- and phenotype data but map data was not provided a JS error is shown on results page and the datatable is malformed.

> DataTables warning: table id=DataTables_Table_0 - Requested unknown parameter '3' for row 0. For more information about this error, please see http://datatables.net/tn/4

### PR DESCRIPTION
**Pull request by kammerer**
Fixes #752 
Fixes #772

### CODE REVIEW

**Comment by Nuanda** [comment]
That's a big one. How fast do you need this one reviewed not to interrupt your flow of work (too much)?

**Comment by kammerer** [comment]
There is no hurry, it does not affect other things I am working on that much.

**Comment by kammerer** [comment]
1. I updated fixtures so that there is a small mismatch between samples, you can use these files to run GWAS: `spec/fixtures/files/gwas-phenotypes.csv`, `spec/fixtures/files/gwas-genotypes-transposed.csv`. You should be able to create the analysis (it was not possible before these changes). The results view should list samples excluded from the analysis.
2. Fixtures `spec/fixtures/files/hapmap/multisample.txt` and `spec/fixtures/hapmap/multisample.gapit.txt` show the difference between hapmap versions. Both header lines should be accepted.
3. If genotype file is identified as hapmap (it starts with "rs" string) but the following headers are wrong you should see validation error when creating the analysis, e.g. "Genotype file contains unexpected column 'allele' (expected 'alleles')".
4. Uploading a text file which is not utf-8 should immediately produce an error message "Invalid file encoding. Expected plain-text or UTF-8 encoded file."  You can use `spec/fixtures/files/text-invalid-encoding.txt`. Blatant fakes may be caught by paperclip's spoofing detection and show different message.

File: app/services/analysis/gwas/csv_normalizer.rb [hunk 1]
```diff
@@ -27,18 +28,29 @@ def call(file, remove_columns: [], remove_rows: [])
...
+
+        tmpfile = create_tmpfile(file.original_filename, normalized_csv)
 
         [:ok, tmpfile]
       end
+
+      private
+
+      def create_tmpfile(original_filename, normalized_csv)
+        NamedTempfile.new(".csv").tap do |tmpfile|
```
[the remark this answers is not in the document]
Comment (kammerer) [reply app/services/analysis/gwas/csv_normalizer.rb]: It is based on `Tempfile` from standard library which is supposed to generate random names.

File: app/services/analysis/gwas/csv_normalizer.rb [hunk 2]
```diff
@@ -27,18 +28,29 @@ def call(file, remove_columns: [], remove_rows: [])
...
             end
           end
         end
 
-        tmpfile = NamedTempfile.new(".csv")
-        tmpfile.original_filename = "#{File.basename(file.original_filename, ".csv")}.normalized.csv"
-        tmpfile << normalized_csv
-        tmpfile.flush
-        tmpfile.rewind
+        return :all_rows_removed if rows_retained < 2
```
[the remark this answers is not in the document]
Comment (kammerer) [reply app/services/analysis/gwas/csv_normalizer.rb]: Yes.

**Comment by Nuanda** [comment]
This is my output to 1.:

```
Following mutations were excluded from the input due to lack of variability: snp1, snp2
Following traits were excluded from the input due to lack of variability: trait1
Following samples were excluded because they were missing from genotype/phenotype data: plant1, plant100, plant2, plant3, plant4, plant5, plant96, plant97, plant98, plant99
```

Correct?

**Comment by Nuanda** [comment]
Re. 2. I used the multisample.vcf but it didn't work with the multisample.txt hapmap file. The ID was missing. The standard phenotype.csv does not work too. What should I use?

**Comment by kammerer** [comment]
Re. 1. Correct.
Re. 2. Try `gwas-phenotypes.csv` with `gwas-genotyes.hapmap` or `gwas-genotypes.gapit.hapmap` (added just now, it only has different headers).

**Comment by Nuanda** [comment]
The output is this:

```
Following mutations were excluded from the input due to lack of variability: rs0000002, rs10904494, rs11252546, rs11591988, rs12775579, rs7909677
Following traits were excluded from the input due to lack of variability: trait1
Following samples were excluded because they were missing from genotype/phenotype data: plant1, plant2, plant3, plant4, plant5
```

Is it correct?

**Comment by kammerer** [comment]
Yes, it is ok.

**Comment by Nuanda** [comment]
All specs pass (apart from the long-broken ones :/). Do you want me to test something else?

**Comment by kammerer** [comment]
No, I don't have anything specific in mind.

**Pull request merged**